### PR TITLE
Eliminate SyntaxArena.default

### PR DESCRIPTION
**Pull request by rintaro**
SyntaxArena.default was a global arena which just leaked. Eliminate it, and make all Syntax factory/modifier API use a new arena.

rdar://101894396

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by jpsim** [comment]
Big +1 from me, this is pretty amazing.

In terms of memory usage, I'm seeing a 2-3x improvement over `main` with #1050, going from 465MB to 162MB in one benchmark.

In terms of time performance, SwiftLint sees a 19% to 47% (!!) overall processing time speedup, and that's considering other processing time not related to SwiftSyntax.

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@jpsim I updated the PR. I don't think this would change the performance much, but could you test it again, just to make sure?

File: Sources/SwiftSyntax/Raw/RawSyntax.swift [hunk 1]
```diff
@@ -219,7 +219,7 @@ extension RawSyntax {
   /// If the syntax tree did not contain a token and thus no trivia could be attached to it, `nil` is returned.
   /// - Parameters:
   ///   - leadingTrivia: The trivia to attach.
-  func withLeadingTrivia(_ leadingTrivia: Trivia) -> RawSyntax? {
+  func withLeadingTrivia(_ leadingTrivia: Trivia, arena: SyntaxArena) -> RawSyntax? {
```
Comment (bnbarham) [inline Sources/SwiftSyntax/Raw/RawSyntax.swift:222]: Is this actually needed? We're always using the one arena for `RawSyntax` anyway aren't we? You also removed these in your other PR, so there'll be a conflict (?)

**Review by bnbarham — APPROVED** [review]

**Review by CodaFi — APPROVED** [review]
💯

**Comment by jpsim** [comment]
I see similar performance with 28813884a1659f3cd2fa61b5b3e5d01e8c25bc34

File: Sources/SwiftSyntax/Raw/RawSyntax.swift [hunk 2]
```diff
@@ -219,7 +219,7 @@ extension RawSyntax {
   /// If the syntax tree did not contain a token and thus no trivia could be attached to it, `nil` is returned.
   /// - Parameters:
   ///   - leadingTrivia: The trivia to attach.
```
Comment (jpsim) [inline Sources/SwiftSyntax/Raw/RawSyntax.swift:221]: Not sure if you care about keeping parameter docstrings up to date, but if so this one’s missing.

**Comment by rintaro** [comment]
@swift-ci Please test

File: Sources/SwiftSyntax/Raw/RawSyntax.swift (hunk 2 above)
Comment (rintaro) [reply Sources/SwiftSyntax/Raw/RawSyntax.swift:221]: Thanks for noticing. Updated.

**Comment by rintaro** [comment]
@swift-ci Please test

**Pull request merged**